% Template for PLoS
% Version 1.0 January 2009
%
% To compile to pdf, run:
% latex plos.template
% bibtex plos.template
% latex plos.template
% latex plos.template
% dvipdf plos.template

\documentclass[12pt]{article}

% amsmath package, useful for mathematical formulas
\usepackage{amsmath}
% amssymb package, useful for mathematical symbols
\usepackage{amssymb}

% graphicx package, useful for including eps and pdf graphics
% include graphics with the command \includegraphics
\usepackage{graphicx}
\usepackage{epstopdf}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

\usepackage{color} 

% Use doublespacing - comment out for single spacing
\usepackage{setspace} 
\doublespacing


% Text layout
\topmargin 0.0cm
\oddsidemargin 0.5cm
\evensidemargin 0.5cm
\textwidth 16cm 
\textheight 21cm

% Bold the 'Figure #' in the caption and separate it with a period
% Captions will be left justified
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}

% Use the PLoS provided bibtex style
\bibliographystyle{plos2009}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother


% Leave date blank
\date{}

\pagestyle{myheadings}
%% ** EDIT HERE **


%% ** EDIT HERE **
%% PLEASE INCLUDE ALL MACROS BELOW

%% END MACROS SECTION

\begin{document}

% Title must be 150 characters or less
\begin{flushleft}
{\Large
\textbf{Computational Identification and Annotation of Key-reactions in Metabolic Pathways of E. coli that Discriminate Different Growth Conditions.}
}
\bigskip
\noindent
% Insert Author names, affiliations and corresponding author email.
\\
Viswanadham Sridhara$^{1}$,
Austin G. Meyer, 
Jeffrey E. Barrick$^{1,2}$,
Pradeep Ravikumar$^{3}$,
Daniel Segre$^{4}$, 
Claus O. Wilke$^{1,5,\ast}$
\\
\bigskip
\bf{1} Center for Computational Biology and Bioinformatics, The University of Texas at Austin, Austin, TX, USA
\\
%\\[3\baselineskip]
\bf{2} Department of Chemistry and Biochemistry, The University of Texas at Austin, Austin, TX, USA
\\
%\\[3ex]
\bf{3} Department of Computer Science, The University of Texas at Austin, Austin, TX, USA
\\
%\\[3ex]
\bf{4} Department of Biology, Boston University, Boston, MA, USA
\\
%\\[3ex]
\bf{5} Section of Integrative Biology, The University of Texas at Austin, Austin, TX, USA
\\
%\\[3ex]
\bigskip
$\ast$ E-mail: wilke@austin.utexas.edu
\end{flushleft}
%\\
% Please keep the abstract between 250 and 300 words
\newpage
\section*{Abstract}
Biochemical information about the enzymes catalyzing reactions in metabolic pathways of microbes is obtained from databases, such as BiGG, BioCyc. This information along with flux balance analysis (FBA) is routinely used in metabolic engineering applications, such as predicting phenotypes given known substrate uptake or gene deletions. In this study, we developed a computational framework in identifying the growth conditions, given the phenotype (simulated flux data). For this, we generated metabolic flux data using FBA on E. coli iAF1250 model for different growth conditions. Using this simulated flux data, we then used model selection algorithm, LASSO, for identifying the key reactions that discriminate different growth conditions. In this study, we used 7 each of carbon/nitrogen sources along with more commonly used experimental media to grow K-12 MG1655 strain. We showed that separately predicting the carbon and nitrogen sources in carbon-nitrogen mixture is better than jointly predicting. On average, 9 features (reactions) were predicted per growth condition. The misclassification rate seems to increase with increasing background noise levels and decreasing training data size. We then mapped these features (reactions) on the E. coli central carbon metabolism to visually show the key genes/metabolites specific to a particular growth condition. For the more commonly used growth media, the models seems to predict at high accuracy too. 

% Please keep the Author Summary between 150 and 200 words
% Use first person. PLoS ONE authors please skip this step. 
% Author Summary not valid for PLoS ONE submissions.   

\section*{Author Summary}
In the last decade, there is gaining popularity for predicting the reaction fluxes in metabolic network using criteria such as maximizing the biomass composition or maximizing the co-factors yield using genome based metabolic models. Moreover, metabolic pathways provide us a framework to integrate diverse kinds of high-throughput data i.e., transcriptomics/proteomics/metabolomics data. Here, we generated metabolic fluxes for different growth conditions and then used machine learning techniques to predict the key-reactions that discriminate different growth conditions. Such automatic identification of key-reactions would in turn also help experimentalists to quantitatively measure the respective metabolites involved (mass-spec based metabolomics), proteins that catalyze the reactions (mass-spec based proteomics) or the genes that encode these enzymes (sequencing based RNA-seq experiments).

\section*{Introduction}
Microbial systems biology is developing at a rapid pace, with advances in sequencing technologies. Genome sequence along with the available annotation can be used to build a biochemical network of metabolic pathways. These metabolic pathways, when represented by mathematical models provide relationship of phenotype to its genotype. Such phenotype prediction given the growth nutrients or the mutant type (gene deletion etc) is generally carried out using Flux Balance Analysis (FBA) \cite{Orthetal2010} on genome scale metabolic  models (GEM) \cite{EdwardsPalsson2000, Karpetal1996, OuzounisKarp2000}. So, given the growth condition, reaction flux vector (phenotype) could be predicted. But can we predict the growth condition, given the simulated flux data? This is the question we are interested in answering as it has many applications. For example, microbes can be engineered in lab for useful purposes as well as deliberate attacks as shown in anthrax mailings. To identify the source of pathogen, mathematical models that can predict the growth condition given the cellular composition of the pathogen would be useful. Hence, in this study, the goal is to identify the reaction fluxes that discriminate the input growth conditions. 

\bigskip
\noindent
Here, we used E. coli iAF1260 GEM model \cite{Feistetal2007} that is comprehensively and qualitatively well annotated by 2 groups for almost 2 decades \cite{EdwardsPalsson2000,Reedetal2003,Feistetal2007,Orthetal2012}. For the current analysis, we used 7 each of carbon and nitrogen sources, to generate simulated flux data. These sources were previously shown to result in growth with this iAF1260 model \cite{Feistetal2007}. Once we have the flux data, we then used machine learning techniques to predict the growth sources. To our knowledge, there were no studies that identified growth sources from flux data.

\bigskip
\noindent
Machine learning algorithms use with computational biology has also been growing, given the high-dimensional nature of the problems in computational biology. In this biological data, generally the number of samples is far less than features. So, linear models cannot be used without reducing the number of covariates.  For example, differential expression of tens of thousands of genes in microarray studies or identifying the SNPs in GWAS studies is a routine task in biomedical research now-a-days. Even though the sequencing technologies are becoming cheaper day-by-day, the number of samples sequenced is still low, compared to the number of features. In such cases, model selection algorithm LASSO \cite{Tibshirani1996} seems to perform well. LASSO methods are used in the past for genomics studies \cite{Wuetal2009}. LASSO seems to perform well with other kinds of data, for example, classifying structural images of brain using MRI data \cite{Casanovaetal2011,Casanovaetal2012,Wangetal2012}.

\bigskip
\noindent
So, in this study, we used LASSO \cite{Friedmanetal2010} along with FBA to answer the following questions: 1. Can we predict the nutrient source on which the microbe is grown, given the simulated flux data? 2. In a mixture of different growth sources, can we predict each growth source sperately or only joint prediction a possibility? 3. Can we make any mechanistic insights into the predicted features (here, reactions in metabolic pathways)? These features (reactions) can later be linked to quanitiatve measurements in targeted OMICS studies.

%\bigskip
%\noindent
%There are many R and MATLAB packages that can be used for LASSO. GLMNET  \cite{Friedmanetal2010} is one such package that can be used for methods based on convex penalties (lasso, ridge-regression, elastic net etc). This package also comes with algorithmic efficiency and speed \cite{Friedmanetal2010}. GLMNET is shown to work on large datasets and with many labels. Given the sparsity of the reaction flux data, we used GLMNET \cite{Friedmanetal2010} for classifying different growth conditions.

% Results and Discussion can be combined.
\section*{Results}

Improving the growth rate or identifying the enzymes key for improving specific metabolite concentrations is a routine task in metabolic engineering. FBA can be regarded as a technique to computationally guide experimentalists, for example, identifying a specific mutant or growth condition that increases a particular engineering target of interest. Here, we used FBA with machine learning techniques to predict growth sources from simulated flux data.

\bigskip
\noindent
1. For FBA, we used COBRA toolbox \cite{Schellenbergeretal2011} with MATLAB.

\bigskip
\noindent
2. For classification, we used GLMNET \cite{Friedmanetal2010} package with R.

\bigskip
\noindent
3. We used E. coli model iAF1260, downloaded from BiGG database \cite{Schellenbergeretal2010}.

\noindent
Below we describe the results obtained in this study.

\subsection*{Simulating metabolic flux datasets:}
The iAF1260 metabolic model of K-12 MG1655 strain was obtained from BiGG database \cite{Schellenbergeretal2010}. This model has 2382 reactions, that included the transport, biochemical and biomass composition reaction. For regression analysis, we removed all the transport reactions in iAF1260 model as these are collinear to the input growth sources, which we are trying to predict from flux data. Other details about this model are provided in methods section.

\bigskip
\noindent
For different input growth conditions that are listed in Table 1, we generated fluxes. To see the effect of noise, we added randomly picked carbon and nitrogen sources from previously known list \cite{Feistetal2007}, in addition to the primary sources we want to predict. Use of noise also allows us to generate many replicates for the same combination of sources, although generating replicates by other ways exists, for example, by finding alternate optima \cite{MahadevanSchilling2003}.

\bigskip
\noindent
We generated 100 replicates of pairwise combinations of 7 carbon and 7 nitrogen sources, i.e., 4900 observations. We then divided the dataset into training and test sets. We trained the models with training data using GLMNET package and used the test set for calculating the misclassification rates. We generated flux data and calculated misclassification rates in the same manner for other commonly used media described in Table 2. Below, we summarize the results.

\bigskip
\subsection*{Noise and training data size effects}
 Contamination in the growth media would generate a different reaction flux vector than without contamination. This would then have an effect on the classification rate of the nutreint sources from simulated flux datasets. To see the effect of varying noise (contamination) levels, we generated simulated flux datasets at different noise levels. We accomplished this by keeping the uptake rate of randomly picked carbon/nitrogen sources at 1\% (0.2 mmol/gDWhr) of the primary sources. For example, 2\% noise means that we randomly picked 2 carbon and 2 nitrogen sources from a set of 174 carbon and 78 nitrogen sources used previously \cite{Feistetal2007}. We then used the cross-validation GLMNET (logistic regression) to pick the model (regression coefficients) that classifies the sources. This cvGLMNET module outputs the misclassification rate on the test set, which we used to present our results.

\bigskip
\noindent
Figure 2 shows that as background carbon and nitrogen sources interference increases, the misclassification rate increases. This is the case for both joint prediction and separate prediction of carbon and nitrogen sources. At 1\% noise, the difference between joint and separate predictions is not a lot, but clearly at 10\% noise, misclassification rate is lower for separate prediction. From this analysis, it looks like separate prediction performs better than joint prediction. This result has 2-fold advantages. 1. The classification models can be built by traning individual sources, making the classification relatively easy. 2. Since the number of experimental observations are generally small, joint prediction has less data to train compared to individual prediction. Even at 10\% noise level with joint prediction, the misclassification rate is 25\% using these models. On the other hand, given 7 carbon and 7 nitrogen sources, just by chance the misclassification rate is expected at 93\%.

\bigskip
\noindent
To make sure the above result has no biases in the uptake amounts, we looked at these uptake distributions for carbon and nitrogen sources. The distributions can be found in supplementary info as S1 figure. Most of the carbon sources uptake seems to be limited by nitrogen and few nitrogen sources seems to have limited carbon for their growth. So, we changed the uptake rates of carbon and nitrogen sources so that there is no limiting factor and re-analyzed by training a new model. These results also suggest that separate prediction works better than joint prediction.

\bigskip
\noindent
Noise or contamination is not the only cause for misclassification with these models. Traning data size has a dominant effect on classification too, as a very small training set would not generate a model that can classify at higher accuracy. To see this effect, we took 3 subsets of original dataset that correspond to 240, 480 and 2400 observations. Using this data, we trained new models with GLMNET package. Our results show that as training data size increases, misclassification decreases. For example, in figure 3, clearly, with a large training data set size of ~2400 observations, the classification is better compared to that of ~480 observations. The low number of observations is more common with real-world experimental datasets and clearly separate prediction outperforms joint prediction with a training data size of ~480 observations.

\bigskip
\noindent
To understand what the models mean, we investigated more into the features that classify both the carbon and nitrogen sources. On average, the number of features for each source seems to be 9. This is relatively small given the total number of reactions (1443). In one study, it has been shown that for a specific growth condition, the total number of reactions that seem to have high-flux are very low \cite{Almaasetal2004}, although the reactions with low-flux can be indicative of the growth source too. To understand this result in a different way, we used a simpler strategy. For each reaction that has a non-zero weight in the model, we knocked-off the reaction flux, trained a new model (separate prediction) and calculated the test accuracy. Except for 3 reactions (predictors), the misclassification rate with the newer models seems to be the same for the rest 126 key-reactions (features predicted for 7 carbon and 7 nitrogen sources). This indicates that alternate models exist, depending on the input set of reaction fluxes.

\bigskip
\noindent
The possible growth nutrients could be many and some of these sources are a different variants of others. For example, maltose is 2 units of glucose. To see how a model classifies a growth source that is not used in training it, we generated flux data using maltose and nitrogen sources that were used above in the study. Since maltose is 2 units of glucose, we expect the model to classify this dataset as glucose (separate prediction), or glucose + nitrogen source (joint prediction). However, the result is quite interesting. Individual prediction seems to classify maltose as glucose more than 85\% of the times and the nitrogen sources were predicted correctly more than 95\% of the times. But when we tried to predict using the joint model, the nitrogen sources were mispredicted most of the times and the carbon source was predicted as D-sorbitol most of the times. 

\subsection*{Mechanistic Insights}
\bigskip
\noindent
To have mechanistic insights into the misclassification caused by noise levels, we drew the heatmap (Figure 4) of the actual source against the predicted sources at 2 noise levels (1\% and 20\%). As seen in figure 4, the results indicate that as noise level increases from 1\% to 20\%, carbon sources get predicted as actate, while N sources get predicted as ammonia most of the times. So, we looked into the reactions that are key in classifying acetate and ammonia sources. Looking at these reactions that are predicted by GLMNET, it seems that they either are part of the TCA cycle or lead to the TCA cycle. This might be due to the objective function used, which is maximizing biomass composition. Generally with any input growth source, the reactions in the pathway that lead to TCA should have some reasonable amount of flux calculated by FBA, if the key is to produce biomass.

\bigskip
\noindent
We mapped the predicted reactions (from separate prediction at 1\% noise level and using 2400 observations for training data size) for each growth source onto the E. coli central metabolism model. We overlayed these reactions on E. coli central metabolism along with the metabolites involved in the reaction. Manual validation of these reactions indicate that indeed some reactions seem specific to growth source, for example glycerol dehydrogenase, glycerol kinase and glycerol-3-phosphate dehydrogenase reactions are specific to "glycerol" source \cite{ZZZZ}. E. coli metabolic map showing predicted reactions for 4 carbon and nitrogen sources is shown in figures 5 and 6. All reactions that are predicted by GLMNET is provided in supplementary information as Table S1.

\bigskip
\subsection*{Large-scale prediction of Carbon/Nitrogen sources}
Since individual prediction is shown to perform better than joint prediction, we did large-scale prediction of the comprehensive list of 174 carbon and 78 nitrogen sources used previously \cite{Feistetal2007}. We used the similar methodology as we used earlier i.e., we generated simulated fluxes for all pair-wise combinations of carbon/nitrogen sources (174*78) for 2 replicates. We used 1 replicate for training the model and then we used the 2nd replicate for testing. The misclassification rate for carbon sources is X\%, while the misclassification rate for nitrogen sources is Y\%.  It should be noted that joint prediction would result in 174*78 classes with 1 replicate for training, while the separate prediction would have 78 replicates for training carbon sources and 174 replicates for training nitrogen sources (pair-wise designs in FBA).

\bigskip
\subsection*{Further validation with other media}
We did similar analysis for the media that are more commonly used in E. coli experiments. The media we used were AB minimal media, ATCC media, Davis Mingioli media and Bochneder defined minimal media. Since the number of different media used in our analysis was 4, we were able to classify at a higher rate even at higher noise levels and lower number of replicates. The misclassification rate with the test set is found to be less than 1\% for noise levels upto 20\%. There was only one feature predicted for each of 3 reactions, except for AB minimal media, for which only $\beta_0$ is non-zero and the rest of regression coefficients are 0. 

\section*{Discussion}
The biochemical network of metabolic pathways, along with FBA, are used for many diverse kinds of applications. For example, a study identified minimal set of reactions that are necessary for growth for a particular growth condition \cite{Burgardetal2001}. There is another study published recently that identified minimal set of nutrients for growth conditions \cite{Ekeretal2013}. Another study by Ibarra et. al., \cite{Ibarraetal2002} looked at the growth of E. coli K-12 MG1655 strain on glycerol for 40 days (~700 generations) and they saw an increase in the growth rate with generations. Other than the above studies, this set of pathways when integrated with other diverse types of data, have many applications as discussed in these 2 reviews \cite{Hydukeetal2013,Oberhardtetal2009}. However, in this study we use the simulated flux data (obtained from different growth conditions) to train mathematical models that can predict the input growth condition.

\bigskip
\noindent
In the current study, we used E. coli metabolic model for the analysis. We used iAF1260 genome-based metabolic model of K-12 MG1655 strain with COBRA toolbox to generate the flux data. We then used GLMNET package to predict different growth conditions, including the media typically used to grow this K-12 strain. Although there is a latest realease of this model \cite{Orthetal2012}, our results should still hold with the new model. One of the reasons we used the previous model is because iAF1260 model is used in many FBA studies and is shown to correlate well with experimental data. We used the constraints of maximizing growth to find a particular reaction flux vector for a given set of growth conditions. Although, the experimental data can be used to refine the constraints \cite{Brandesetal2012}, we used the default lower and upper bounds in the model. We generated replicates for the primary growth sources of interest, by solving for reaction flux vector at different varying background levels of growth sources (contamination). We then used the simulated flux data to train mathematical models to predict the input growth sources. 

\bigskip
\noindent
Our results indicate that separate prediction of the mixture of growth sources performs well than joint prediction at a noise level, for a given size of training data. This result is interesting, as we intially thought there would be lot of cross-talk between the carbon and nitrogen sources. This result is advantageous, as the model trained on known growth sources would predict those sources, when present, at a high confidence level when tested on species grown on unknown nutrients. Although the simulated flux data is a good first step to train the models, ideally we would like to integrate these models to use the experimental data. For example, combining flux analysis data with phosphoproteomics data to deduce functionality of enzymes is described in this study \cite{Oliveiraetal2012}. An interesting study to understand the adaptation of yeast metabolism to the growth conditions is studied, using enzymes in metabolic model pathways to validate the quantitative proteomics data \cite{Costenobleetal2011}.
 
%\bigskip
%\noindent
%
%\bigskip
%\noindent
%We showed that given simulated metabolic flux data, growth conditions can be predicted at good accuracy. The noise levels and training data seem to have effect on prediction rate. Given the same number of observations in training set, separate prediction seems to perform well than joint prediction. Few signature reactions for a specific growth source were predicted in our study, which validates the methodology used in this study.

\bigskip
\noindent
To our knowledge, LASSO methods are not used in the analysis of metabolic pathways. However there are studies that used LASSO on other biochemical networks such as gene regulatory networks \cite{Menendezetal2010}. In machine learning algorithms, there are other regression methods that use regularization techniques other than LASSO. Graphical LASSO \cite{Friedmanetal2008} and Ising Markov Random Field models \cite{Ravikumaretal2010} are also used to complement with LASSO. We would like to explore these and other models to see if we can improve the classification rate.

% You may title this section "Methods" or "Models". 
% "Models" is not a valid title for PLoS ONE authors. However, PLoS ONE
% authors may use "Analysis" 

\section*{Materials and Methods}
We used MATLAB and R for this study. For flux balance analysis, we used COBRA toolbox \cite{Schellenbergeretal2011} with MATLAB and for multinomial regression, we used GLMNET package \cite{Friedmanetal2010} with R. The methods are described in detail below.

\subsection*{Flux Balance Analysis:} 
Biochemical network represented mathematically (metabolites and reactions) is used in flux balance analysis. In FBA, the steady-state solution for reaction fluxes is calculated. S(m,n) is the stoichiometric matrix for "m" metabolites and "n" reactions and is represented as S hereafter. The other variables used are v and x, where v is a vector of reaction fluxes for all the reactions involved and x is the concentration of the metabolites. In steady-state, the rate of change of metabolite concentrations is 0. So, we can formulate the above problem with the set of equations, as described below:

\begin{align}
dx/dt = Sv_{i}, \\
Sv_{i}=0.
\end{align}

The constraints that are typically used are: 
\begin{align}
\alpha_{i}<v_{i}<\beta_{i}
\end{align}
where $\alpha$ and $\beta$ are lower and upper bounds of the reaction fluxes.

Hence, we solve this set of linear equations with interested constraints by linear programming.
\begin{align}
{max.} c^Tv, s. t Sv_{i}=0.
\end{align}
where $c^Tv$ represents the biomass composition reaction.

\bigskip
\noindent
Generally in these metabolic networks, the number of reactions are more than number of metabolites resulting in multiple solutions.  However in metabolic engineering applications, a certain objective function of interest is optimized, for example here, we used a constraint to maximize biomass composition. This resulted in uniqe solution. 

\bigskip
\noindent
Equation (4) can be solved for different input growth conditions i.e., by changing lower bounds of transport reactions that uptake a specific nutrient. To create replicates, for each design, we introduced varying noise levels. Our motivation to introduce noise is 2-fold. First, we would like to find alternate optimal solutions using FBA methods so that the final prediction results still hold for varying noise levels. The second reason is to simulate replicates for a particular growth condition to train the mathematical models. Flux variability analysis can also be used for finding alternate optimal solutions, but apart from generating alternate optimal solutions and generating replicates, we can also introduce noise and find its effect in key-reaction prediction using the method described above.

\subsection*{E. coli model:} 
From the BiGG database, we downloaded the iAF1260 model, as it has shown to be used rigorously in various studies involving metabolic engineering and seems to agree well with the experimental data. Generally these models are stored in SBML format, which is becoming a common format for systems biology related models i.e., signaling pathways, gene regulatory networks, metabolic pathways etc \cite{Huckaetal2003}. In the current iAF1260 model, there are 2382 reactions, 1668 metabolites. The biomass composition reaction is also included in the model. Except for the input growth sources (Carbon and nitrogen sources used in this study) used, we did not change any defaults that are used in this model. The upper bounds on 2377 reactions is set to 1000 mmol/gDWhr, i.e., there is no limit on the production of metabolites involved in these reactions. But for 5 reactions, i.e., ATPM, CAT, FHL, SPODM, SPODMpp, the upper bound was set to 50 mmol/gDWhr. On the other hand the lower bound for almost 1800 reactions is set to 0 mmol/gDWhr, which means these reactions cannot uptake any metabolites from the media. For ATPM reaction, the lower bound is set to be the same as upper bound at 8.39 mmol/gDWhr. For the rest, the lower bounds were set to -1000 mmol/gDWhr, except for glucose and oxygen. We did not change the oxygen uptake rate (-18.5mmol/gDWhr), but we set the lower bounds of glucose and ammonia to zero. If we used glucose/ammonia as growth sources, we then set the lower bounds of these sources accordingly.

\subsection*{Growth conditions:} 
We picked the growth conditions manually from \cite{Feistetal2007} that seemed more common in experiments. In our study, we used pairwise combinations of 7 carbon sources, and 7 nitrogen sources. 7 carbon sources when used alone did not result in any growth. On the other hand, the nitrogen sources except ammonia contributed to E. coli biomass composition that is non-zero. These carbon and nitrogen sources are listed in Table I. Depending on the input growth, we set the lower bound of that particular exchange reaction to -20 mmol/gDWhr. This lower bound of -20 mmol/gDWhr is previously used as reasonable uptake amount in many studies \cite{Feistetal2007}. For 49 pair-wise combinations of the sources, we generated 100 replicates of data. Apart from these growth conditions, we also used 4 growth media, generally used in E. coli K-12 MG1655 experiments as cited in EcoCyc database [Cite URL of EcoCyc].
\bigskip
\noindent
We changed the uptake rate of nitrogen source to -1000 mmol/gDW/hr, keeping the carbon source at -20mmol/gDW/hr to make sure carbon sources don't lack nitrogen source. We also repeated the analysis keeping the carbon source uptake at -1000 mmol/gDW/hr and nitrogen source uptake at -20mmol/g/DW/hr. 


\subsection*{Background noise levels:} 
To generate replicates and may be alternate optimal solutions using these conditions, we incorporated different background noise levels. For this, we used a subset of the 174 carbon and 78 nitrogen sources, previously used in Feist et. al \cite{Feistetal2007}.  We used different background noise levels, ranging from 1\% to 20\%. For example, if we want to set 5\% noise level, we randomly picked 5 Carbon and 5 nitrogen sources and set their lower bounds to -0.2 mmol/gDW. Please note that we generated the flux data for a pairwise combination of 1 carbon and 1 nitrogen source along with the background noise as described here. 
%[Since we are using different noise sources, we used a minimum threshold for biomass composition (0.558)].

\bigskip
\noindent
For each noise level, we used half of the dataset as test set. We used subsets of the remaining half as training (i.e, 240,480,2400 observations). On the training sets we did 3-fold cross validation. We used cross-validation in GLMNET package for model selection. Model selection means picking the regression coefficients at the lambda value that had the lowest misclassification rate with 3-fold cross-validation. We then used this model to calculate the misclassification rate on the test set. We repeated this step to calculate the misclassification rates at different noise levels (1\%,5\%,10\%) and different training data sizes (240,480,2400 observations).

\subsection*{Regression based on regularization:} 
We used GLMNET package with R. We used 3 fold cross-validation. From the flux balance analysis, the observations we generated for different noise levels were divided into 2- halves, one for training and the other for test set. We did training and 3-fold cross validation to pick the lambda that has the lowest misclassification rate using the GLMNET. We did this for both joint prediction as well as the separate prediction and then making a joint call.

\bigskip
\noindent
1. Separately:
\noindent

i) Take data set
\noindent

ii) Train model to predict C sources
\noindent

iii) Train model to predict N sources
\noindent

iv) Predict C and N separately and calculate joint prediction accuracy.

\bigskip
\noindent
2. Jointly
\noindent

i) Take data set
\noindent

ii) Train model to predict C and N jointly
\noindent

iii) Predict C and N jointly and calculate prediction accuracy.

\bigskip
\noindent
Since the cross-validation accuracy cannot be used to compare the GLMNET results for separately predicting to that of joint prediction, we used the test set to determine the prediction accuracy.

\bigskip
\noindent
Below are the equations used in multinomial regression setting. If Y is a categorical response variable with "m" levels (m>2), and x is a predictor, this will result in
\begin{align}
log P_r(Y=l/x)/P_r(Y=m/x) = \beta_0l+x^T\beta_l, l=1,2,...,m-1.
\end{align}

From this, we can model
\begin{align}
P_r(Y=l/x) = e^( \beta_0l+x^T\beta_l)/\sum\limits_{k=1}^m e^( \beta_0l+x^T\beta_l).
\end{align}

The above model can be fit by maximizing the penalized log-likelihood, as explained in detail elsewhere \cite{Friedmanetal2008}
\begin{align}
{max} 1/N \sum\limits_{i=1}^N log p_gi(x_i) - \lambda\sum\limits_{l=1}^m P_\alpha(\beta_l)
\end{align}


\noindent
Here $\beta$ are the regression coefficients and N is the total number of observations. However, in LASSO, there is a tuning constant $\lambda$ that puts the strength on the penalty introduced to achieve sparsity. We direct the reader to Friedman et. al., \cite{Friedmanetal2008} for further details.

\bigskip
\noindent
We used GLMNET package with R, instead of MATLAB as the supercomputing cluster at TACC has R installed on it and we were able to easily add the GLMNET package to it.

\subsection*{Mechanistic Insights:} 
Mechanistic insights of the results obtained from multinomial regression can be obtained by understanding the role of features that are predicted for each growth condition. For this, we used E. coli map downloaded from BiGG database. For overlaying reactions (features) onto E. coli central metabolism map, we used modules in COBRA toolbox. We deleted 4 reactions in the map that seemed inconsistent with the E. coli model used. The predictors from GLMNET are highlighted with a different color, along with the metabolites involved in these reactions.


% Do NOT remove this, even if you are not including acknowledgments
\section*{Acknowledgments}
We would like to thank Segre lab members at Boston University, Marx lab members at Harvard University for useful discussions on flux balance analysis. We would also like to thank BCG and TACC at UT for resources. VS would like to thank Piyush Rai for useful discussions on multinomial classification.

\section*{Author Contributions}
Conceived and designed the experiments: V.S. and C.O.W. Performed the experiments: V.S. Analyzed the data: V.S and C.O.W. Wrote the paper: V.S,  A.G.M, J.E.B, P.R, D.S. and C.O.W.

%\section*{References}
% The bibtex filename
\bibliography{FBAGLMNETBibliography}


\newpage
\section*{Tables}
\begin{table}[!ht]
\caption{
\bf{Carbon and nitrogen sources used in the study.}}
%\includegraphics[width=5in]{Figures/TableSources1.png}
\begin{tabular}{|c|c|}
\hline \hline
Carbon Sources & Nitrogen Sources \\
\hline
D-glucose & Ammonia \\
Pyruvate & Adenine \\
Glycerol & Cytidine \\
Acetate & Putrescine \\
D-ribose & L-glycine \\
D-fructose & L-alanine \\
D-sorbitol & L-glutamine \\
\hline
\end{tabular}
\begin{flushleft} 7 carbon and 7 nitrogen sources are used in this study.
\end{flushleft}
\label{tab:label}
 \end{table}


\begin{table}[!ht]
\caption{
\bf{Growth media for K-12 MG1655.}}
%\includegraphics[width=3in]{Figures/TableSources2.png}
\begin{tabular}{|c|c|c|}
\hline \hline
Other media & Composition \\
\hline
AB medium &  D-glucose/Ammonia\\
ATCC medium 57 &  Glycerol/L-lysine/Ammonia\\
Bochner medium &  Pyruvate/Ammonia\\
Davis Mingioli medium & D-glucose/Citrate/Ammonia\\
\hline
\end{tabular}
\begin{flushleft}4 commonly used growth media for E. coli MG1655 strain that were picked from EcoCyc website were also used in this study.
\end{flushleft}
\label{tab:label}
 \end{table}

\bigskip
\section*{Figure Legends}
\begin{figure}[!ht]
\begin{center}
\includegraphics[width=4.5in]{Figures/flowchart_new.pdf}
\end{center}
\caption{
{\bf Flowchart describing methodology used in this study.} We obtained E. coli model and map from BiGG database. The key steps involved are Flux Balance Analysis and multinomial classification routines.
}
\label{Figure_label}
\end{figure}

\begin{figure}[!ht]
\begin{center}
\includegraphics[width=6in]{Figures/VaryingData.png}
\end{center}
\caption{
{\bf Misclassification rate for different replicate sizes.} This plot shows that as training data size increases, the misclassification rate increases. This is tested for 2 different noise levels (1\% and 10\%). 
}
\label{Figure_label}
\end{figure}

\begin{figure}[!ht]
\begin{center}
\includegraphics[width=6in]{Figures/VaryingNoise.png}
\end{center}
\caption{
{\bf Misclassification rate at different noise levels.} This plot shows that misclassification rate increases as noise increases in FBA models. This is tested for 2 different replicate sizes (480 and 2400 observations). 
}
\label{Figure_label}
\end{figure}

\begin{figure}[!ht]
\begin{center}
\includegraphics[width=6in]{Figures/heatmapNEWEST.jpg}
\end{center}
\caption{
{\bf Heat maps with actual sources as columns and predicted ones in rows.}  At 20\% noise, most C sources are predicted to be acetate and most N sources are predicted to be ammonia.
}
\label{Figure_label}
\end{figure}

\begin{figure}[!ht]
\begin{center}
\hspace*{-3.5cm}
\includegraphics[width=9in]{Figures/CarbonSources.jpg}
\end{center}
\caption{
{\bf Key-reactions that discriminate Carbon sources.}  The key-reactions identified by GLMNET package were mapped onto E. coli central metabolism to visually show the differences between different growth conditions. Out of 7 carbon sources, here we show 4 carbon sources and the key-reactions.
}
\label{Figure_label}
\end{figure}

\begin{figure}[!ht]
\begin{center}
\hspace*{-3.25cm}
\includegraphics[width=8.5in]{Figures/NitrogenSources.jpg}
\end{center}
\caption{
{\bf Key-reactions that discriminate Nitrogen sources.}  The key-reactions identified by GLMNET package were mapped onto E. coli central metabolism to visually show the differences between different growth conditions. Out of 7 nitrogen sources, here we show 4 nitrogen sources and the key-reactions.
}
\label{Figure_label}
\end{figure}

%\begin{figure}[!ht]
%\begin{center}
%\hspace*{-3.5cm}
%\includegraphics[width=9in]{Figures/MediaSources.png}
%\end{center}
%\caption{
%{\bf Key-reactions that discriminate other media.}  The key-reactions identified by GLMNET package were mapped onto E. coli central metabolism to visually show the differences between different growth conditions. Here, the growth medium used are generally used for K-12 MG1655 strain.
%}
%\label{Figure_label}
%\end{figure}

\end{document}
